sdl2promela/translator.py

```diff
@@ -296,36 +386,498 @@ def __generate_statement(
+    sdlmodel.Sequence,
+    type,
+)
+def __generate_assignment(
+    sdl_model: sdlmodel.Model,
+    left: Union[
+        sdlmodel.VariableReference, sdlmodel.ArrayAccess, sdlmodel.MemberAccess
+    ],
+    right: sdlmodel.Sequence,
+    left_type: type,
+) -> List[promelamodel.Statement]:
+    finalType = resolve_asn1_type(sdl_model.types, left_type)
+    if finalType.kind != "SequenceType":
+        raise Exception(
+            f"Invalid assignment: {finalType.CName} does not accept SEQUENCE value"
+        )
+
+    statements: List[promelamodel.Statement] = []
+
+    for name, dataType in finalType.Children.items():
+        is_optional = dataType.Optional == "True"
+        if not is_optional and name not in right.elements:
+            raise Exception(
+                f"Invalid assignment to {finalType.CName}: missing required member {name}"
+            )
+
+        if name in right.elements:
+            statements.extend(
+                __generate_assignment(
+                    sdl_model,
+                    sdlmodel.MemberAccess(left, sdlmodel.VariableReference(name)),
+                    right.elements[name],
+                    dataType.type,
+                )
+            )
+
+        if is_optional:
+            if name in right.elements:
+                value = promelamodel.IntegerValue(1)
+            else:
+                value = promelamodel.IntegerValue(0)
+            statements.append(
+                AssignmentBuilder()
+                .withTarget(
+                    MemberAccessBuilder()
+                    .withUtypeReference(
+                        MemberAccessBuilder()
+                        .withUtypeReference(
+                            __generate_variable_name(sdl_model, left, True)
+                        )
+                        .withMember(VariableReferenceBuilder("exist").build())
+                        .build()
+                    )
+                    .withMember(VariableReferenceBuilder(name).build())
+                    .build()
+                )
+                .withSource(value)
+                .build()
+            )
+
+    return statements
+
+
+@dispatch(
+    sdlmodel.Model,
+    (
+        sdlmodel.VariableReference,
+        sdlmodel.ArrayAccess,
+        sdlmodel.MemberAccess,
+    ),
+    sdlmodel.SequenceOf,
+    type,
+)
+def __generate_assignment(
+    sdl_model: sdlmodel.Model,
+    left: Union[
+        sdlmodel.VariableReference, sdlmodel.ArrayAccess, sdlmodel.MemberAccess
+    ],
+    right: sdlmodel.SequenceOf,
+    left_type: type,
+) -> List[promelamodel.Statement]:
+    finalType = resolve_asn1_type(sdl_model.types, left_type)
+    if finalType.kind != "SequenceOfType":
+        raise Exception(
+            f"Invalid assignment: {finalType.CName} does not accept SEQUENCE OF value"
+        )
+
+    statements: List[promelamodel.Statement] = []
+
+    length = len(right.elements)
+
+    if length < int(finalType.Min) or length > int(finalType.Max):
```

Comment:
This also seems repeated a lot... consider extracting a helper function, validate/check length

Reply:
Ok, this is a separated function now.